src/inc/til/spsc.h

```diff
@@ -0,0 +1,530 @@
+    }
+
+    template<typename T>
+    struct sender
+    {
+        explicit sender(details::arc<T>* arc) noexcept :
+            _arc(arc) {}
+
+        sender<T>(const sender<T>&) = delete;
+        sender<T>& operator=(const sender<T>&) = delete;
+
+        sender(sender<T>&& other) noexcept
+        {
+            drop();
+            _arc = std::exchange(other._arc, nullptr);
+        }
+
+        sender<T>& operator=(sender<T>&& other) noexcept
+        {
+            drop();
+            _arc = std::exchange(other._arc, nullptr);
+        }
+
+        ~sender()
+        {
+            drop();
+        }
+
+        // emplace constructs an item in-place at the end of the queue.
+        // It returns true, if the item was successfully placed within the queue.
+        // The return value will be false, if the receiver is gone.
+        template<typename... Args>
+        bool emplace(Args&&... args) const
+        {
+            auto acquisition = _arc->producer_acquire(1);
+            if (!acquisition)
+            {
+                return false;
+            }
+
+            auto data = _arc->data();
+            auto begin = data + acquisition.begin;
+            new (begin) T(std::forward<Args>(args)...);
+
+            _arc->producer_release(acquisition);
+            return true;
+        }
+
+        // move_n moves count items from the input iterator in into the queue.
+        // The resulting iterator is returned as the first pair field.
+        // The second pair field will be false if the receiver is gone.
+        template<typename InputIt>
+        std::pair<InputIt, bool> move_n(InputIt in, size_t count) const
+        {
+            details::validate_size(count);
+
+            auto data = _arc->data();
+            auto ok = true;
+
+            while (count != 0)
+            {
+                auto acquisition = _arc->producer_acquire(size_type(count));
+                if (!acquisition)
+                {
+                    ok = false;
+                    break;
+                }
+
+                auto begin = data + acquisition.begin;
+                auto got = acquisition.end - acquisition.begin;
+                in = std::uninitialized_move_n(in, got, begin).first;
+                count -= got;
+
+                _arc->producer_release(acquisition);
+            }
+
+            return { in, ok };
+        }
+
+    private:
+        void drop()
+        {
+            if (_arc)
+            {
+                _arc->drop_producer();
+            }
+        }
+
+        details::arc<T>* _arc = nullptr;
+    };
+
+    template<typename T>
+    struct receiver
+    {
+        explicit receiver(details::arc<T>* arc) noexcept :
+            _arc(arc) {}
+
+        receiver<T>(const receiver<T>&) = delete;
+        receiver<T>& operator=(const receiver<T>&) = delete;
+
+        receiver(receiver<T>&& other) noexcept
+        {
+            drop();
+            _arc = std::exchange(other._arc, nullptr);

```

Comment:
So, to clarify, if I try to copy construct the reciever, 
```c++
auto [tx, rx_a] = channel...;
auto rx_b = rx_a;
```

and I try to `pop` from `rx_a`, that'll throw? return `nullopt`? Return the next value?

Reply:
The copy constructor is deleted. You can only move the two handles ("sender" / "receiver").